src/dbnode/persist/fs/write.go

```diff
@@ -509,14 +479,62 @@ func (w *writer) writeIndexFileContents(
 			w.indexEntries[i].indexFileOffset = offset
 		}
 
-		offset += int64(len(data))
+		offset, err = w.writeIndex(id, tagsIter, tagsEncoder, entry, offset)
+		if err != nil {
+			return err
+		}
 
 		prevID = id
 	}
 
 	return nil
 }
 
+func (w *writer) writeIndex(
+	id []byte,
+	tagsIter ident.TagIterator,
+	tagsEncoder serialize.TagEncoder,
+	entry indexEntry,
+	offset int64,
+) (int64, error) {
+	var encodedTags []byte
+	if numTags := tagsIter.Remaining(); numTags > 0 {
+		tagsEncoder.Reset()
+		if err := tagsEncoder.Encode(tagsIter); err != nil {
+			return offset, err
+		}
+
+		encodedTagsData, ok := tagsEncoder.Data()
+		if !ok {
+			return offset, errWriterEncodeTagsDataNotAccessible
+		}
+
+		encodedTags = encodedTagsData.Bytes()
+	}
+
+	e := schema.IndexEntry{
+		Index:        entry.index,
+		ID:           id,
+		Size:         int64(entry.size),
+		Offset:       entry.dataFileOffset,
+		DataChecksum: int64(entry.dataChecksum),
+		EncodedTags:  encodedTags,
+	}
+
+	w.encoder.Reset()
+	if err := w.encoder.EncodeIndexEntry(e); err != nil {
+		return offset, err
+	}
+
+	data := w.encoder.Bytes()
+	if _, err := w.indexFdWithDigest.Write(data); err != nil {
+		return offset, err
+	}
+
+	offset += int64(len(data))
+	return offset, nil
```

Comment:
Can this instead return `int64(len(data)), nil`  and not have an offset passed in? I think injecting the current offset and updating it here makes the functionality a little more difficult to reason about than the caller explicitly appending the length of the written data